File: go/grammar.go
package gencmu

// A stage's grammar: its documents stitched into one set of rules and
// directives (engine §2).
type stageGrammar struct {
	name          string
	rules         []*sRule
	byName        map[string]*sRule
	lean          string // "greedy" or "lazy"
	elisionOnly   bool
	elidable      map[string]bool
	freeModifiers string
	changes       []stitchChange
}

// stitchChange records a rule a later document replaced or extended.
type stitchChange struct {
	rule, document, op string
}

type sRule struct {
	name string
	alts []*sAlt
	doc  string
	at   [2]int
}

// sAlt is an alternative with the clauses of the rule statement that wrote
// it: an extension's alternatives keep the extension's own clauses.
type sAlt struct {
	alt      *domAlt
	ruleTags *domTerm
	emit     *domEmit
	conds    []*domCond
	doc      string
	at       [2]int
}

type docDOM struct {
	path string
	dom  *domDoc
}

func isTerminalName(name string) bool {
	return name != "" && name[0] >= 'A' && name[0] <= 'Z'
}

func stitch(stageName string, docs []docDOM) (*stageGrammar, *Error) {
	g := &stageGrammar{name: stageName, byName: map[string]*sRule{}, elidable: map[string]bool{}}
	fail := func(doc string, at [2]int, format string, args ...any) *Error {
		e := grammarError(doc, at, format, args...)
		e.Stage = stageName
		return e
	}
	var ambiguity []*domDirective
	freeCount := 0
	for _, d := range docs {
		defined := map[string]bool{}
		for _, r := range d.dom.Rules {
			alts := make([]*sAlt, len(r.Alternatives))
			for i, a := range r.Alternatives {
				alts[i] = &sAlt{alt: a, ruleTags: r.Tags, emit: r.Emit, conds: r.Conditions, doc: d.path, at: r.At}
			}
			existing := g.byName[r.Name]
			switch r.Op {
			case "define":
				if defined[r.Name] {
					return nil, fail(d.path, r.At, "%s is defined twice with ≔ in one document", r.Name)
				}
				defined[r.Name] = true
				if existing != nil {
					g.changes = append(g.changes, stitchChange{r.Name, d.path, "replace"})
					existing.alts, existing.doc, existing.at = alts, d.path, r.At
				} else {
					nr := &sRule{name: r.Name, alts: alts, doc: d.path, at: r.At}
					g.rules = append(g.rules, nr)
					g.byName[r.Name] = nr
				}
			case "extend":
				if existing == nil {
					return nil, fail(d.path, r.At, "%s |≔ extends a rule not defined before it", r.Name)
				}
				g.changes = append(g.changes, stitchChange{r.Name, d.path, "extend"})
				existing.alts = append(existing.alts, alts...)
			default:
				return nil, fail(d.path, r.At, "unknown rule operator %q", r.Op)
			}
		}
		for _, dir := range d.dom.Directives {
			switch dir.Name {
			case "ambiguity-resolution":
				ambiguity = append(ambiguity, dir)
				if len(ambiguity) > 1 {
					return nil, fail(d.path, dir.At, "stage %s has more than one %%ambiguity-resolution", stageName)
				}
				args := dir.Args
				if len(args) == 0 || len(args) > 2 || (args[0] != "greedy" && args[0] != "lazy") || (len(args) == 2 && args[1] != "elision-only") {
					return nil, fail(d.path, dir.At, "%%ambiguity-resolution takes greedy or lazy, optionally followed by elision-only")
				}
				g.lean = args[0]
				g.elisionOnly = len(args) == 2
			case "elidable":
				for _, a := range dir.Args {
					g.elidable[a] = true
				}
			case "free-modifiers":
				freeCount++
				if freeCount > 1 {
					return nil, fail(d.path, dir.At, "stage %s has more than one %%free-modifiers", stageName)
				}
				if len(dir.Args) != 1 {
					return nil, fail(d.path, dir.At, "%%free-modifiers takes one rule name")
				}
				g.freeModifiers = dir.Args[0]
			default:
				return nil, fail(d.path, dir.At, "unknown directive %%%s", dir.Name)
			}
		}
	}
	if len(ambiguity) == 0 {
		e := &Error{Kind: "grammar", Stage: stageName, Message: "stage " + stageName + " has no %ambiguity-resolution"}
		if len(docs) > 0 {
			e.Document = docs[0].path
		}
		return nil, e
	}
	if g.freeModifiers != "" && g.byName[g.freeModifiers] == nil {
		return nil, fail(docs[0].path, [2]int{}, "%%free-modifiers names %s, which is not a rule of the stage", g.freeModifiers)
	}
	if g.byName["text"] == nil {
		e := &Error{Kind: "grammar", Stage: stageName, Message: "stage " + stageName + " defines no start rule text"}
		if len(docs) > 0 {
			e.Document = docs[0].path
		}
		return nil, e
	}
	for _, r := range g.rules {
		for _, a := range r.alts {
			if err := g.checkAlt(a); err != nil {
				err.Stage = stageName
				return nil, err
			}
		}
	}
	return g, nil
}

// checkAlt checks what the notation's grammar cannot state: every rule named
// is defined, # has a declaration, and captures stand at the top level.
func (g *stageGrammar) checkAlt(a *sAlt) *Error {
	fail := func(format string, args ...any) *Error { return grammarError(a.doc, a.at, format, args...) }
	captures := map[string]bool{}
	var walk func(e *domExpr, top bool) *Error
	walk = func(e *domExpr, top bool) *Error {
		switch e.Kind {
		case exRef:
			if !isTerminalName(e.Name) && g.byName[e.Name] == nil {
				return fail("%s is not a rule of stage %s", e.Name, g.name)
			}
		case exHash:
			if g.freeModifiers == "" {
				return fail("# is used, but the stage has no %%free-modifiers")
			}
		case exCapture:
			if !top {
				return fail("a capture must stand at the top level of an alternative")
			}
			if captures[e.Name] {
				return fail("$%s is captured twice in one alternative", e.Name)
			}
			captures[e.Name] = true
			if len(captures) > 4 {
				return fail("an alternative has at most four captures")
			}
			if e.Inner.Kind != exRef && e.Inner.Kind != exTerminal {
				return fail("a capture wraps a single symbol")
			}
			return walk(e.Inner, false)
		case exSeq:
			for _, it := range e.Items {
				if err := walk(it, top); err != nil {
					return err
				}
			}
		case exChoice, exAnd:
			for _, it := range e.Items {
				if err := walk(it, false); err != nil {
					return err
				}
			}
		case exOptional, exRepeat:
			return walk(e.Inner, false)
		}
		return nil
	}
	if err := walk(a.alt.Expr, true); err != nil {
		return err
	}
	var checkTerm func(t *domTerm) *Error
	checkTerm = func(t *domTerm) *Error {
		if t == nil {
			return nil
		}
		if t.Kind == tmRule && g.byName[t.Str] == nil {
			return fail("%s is not a rule of stage %s", t.Str, g.name)
		}
		for _, it := range t.Items {
			if err := checkTerm(it); err != nil {
				return err
			}
		}
		return nil
	}
	var checkCond func(c *domCond) *Error
	checkCond = func(c *domCond) *Error {
		switch c.Kind {
		case cdCompare:
			if err := checkTerm(c.Left); err != nil {
				return err
			}
			return checkTerm(c.Right)
		case cdMatches:
			if g.byName[c.Rule] == nil {
				return fail("%s is not a rule of stage %s", c.Rule, g.name)
			}
			return checkTerm(c.Span)
		case cdNot:
			return checkCond(c.Inner)
		case cdAny:
			for _, it := range c.Items {
				if err := checkCond(it); err != nil {
					return err
				}
			}
		}
		return nil
	}
	for _, t := range []*domTerm{a.alt.Tags, a.ruleTags} {
		if err := checkTerm(t); err != nil {
			return err
		}
	}
	for _, c := range a.conds {
		if err := checkCond(c); err != nil {
			return err
		}
	}
	if a.emit != nil {
		for _, it := range a.emit.Items {
			if err := checkTerm(it.Tags); err != nil {
				return err
			}
		}
	}
	return nil
}

// termCaptures lists the captures a term mentions.
func termCaptures(t *domTerm, into map[string]bool) {
	if t == nil {
		return
	}
	if t.Kind == tmCapture {
		into[t.Str] = true
	}
	for _, it := range t.Items {
		termCaptures(it, into)
	}
}

func condCaptures(c *domCond, into map[string]bool) {
	switch c.Kind {
	case cdCompare:
		termCaptures(c.Left, into)
		termCaptures(c.Right, into)
	case cdMatches:
		termCaptures(c.Span, into)
	case cdNot:
		condCaptures(c.Inner, into)
	case cdAny:
		for _, it := range c.Items {
			condCaptures(it, into)
		}
	}
}

File: lib/rust/src/grammar.rs
//! Stitching a stage's documents into one grammar (engine §2), with the
//! checks that need no features: directives, definitions, references,
//! captures and terms.

use std::collections::HashMap;
use std::sync::Arc;

use crate::dom::{Alternative, Arg, Cond, Dom, EmitItem, Expr, Op, Term};
use crate::error::Error;

/// How a stage chooses among parses (engine §6): the lean of rule 2, or,
/// for the `elision-only` check (§7), rule 1 alone.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub(crate) enum Lean {
    Greedy,
    Lazy,
    TagsOnly,
}

/// An alternative after stitching, with the clauses of the rule that
/// contributed it.
#[derive(Debug, Clone)]
pub(crate) struct StitchedAlternative {
    pub alternative: Alternative,
    pub rule_tags: Option<Term>,
    pub emit: Option<Vec<EmitItem>>,
    pub conditions: Vec<Cond>,
    pub verbatim: bool,
    pub document: Arc<str>,
    pub at: (usize, usize),
}

#[derive(Debug, Clone)]
pub(crate) struct StitchedRule {
    pub name: String,
    pub alternatives: Vec<StitchedAlternative>,
    pub document: Arc<str>,
}

/// One replacement or extension the loader recorded (engine §2).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Change {
    /// The stage whose grammar changed.
    pub stage: String,
    /// The rule that was replaced or extended.
    pub rule: String,
    /// True for an extension (`%extend-rule`), false for a replacement
    /// (`%redefine-rule`).
    pub extension: bool,
    /// The document that made the change.
    pub document: String,
}

#[derive(Debug, Clone)]
pub(crate) struct StageGrammar {
    pub name: String,
    pub rules: Vec<StitchedRule>,
    pub index: HashMap<String, usize>,
    pub lean: Lean,
    pub elision_only: bool,
    /// Whether an elided terminator is forbidden where its constituent
    /// could have been longer (engine §4).
    pub maximal: bool,
    pub elidable: Vec<String>,
    pub changes: Vec<Change>,
}

/// The most items an `&` may have (engine §3.2): its expansions are every
/// non-empty subsequence, so more would be too many to lower.
pub(crate) const MAX_AND: usize = 16;

pub(crate) fn is_terminal_name(name: &str) -> bool {
    name.chars().next().is_some_and(|c| c.is_ascii_uppercase())
}

fn located(message: String, document: &str, at: (usize, usize)) -> Error {
    Error::grammar(message).in_document(document).at(at.0, at.1)
}

/// Stitches the documents of one stage, in order.
pub(crate) fn stitch(stage: &str, documents: &[(Arc<str>, Arc<Dom>)]) -> Result<StageGrammar, Error> {
    let mut grammar = StageGrammar {
        name: stage.to_string(),
        rules: Vec::new(),
        index: HashMap::new(),
        lean: Lean::Greedy,
        elision_only: false,
        maximal: false,
        elidable: Vec::new(),
        changes: Vec::new(),
    };
    let mut resolution: Option<(Arc<str>, (usize, usize))> = None;
    for (document, dom) in documents {
        // The rules this document defined or redefined (engine §2).
        let mut defined_here: HashMap<&str, ()> = HashMap::new();
        for rule in &dom.rules {
            let alternatives: Vec<StitchedAlternative> = rule
                .alternatives
                .iter()
                .map(|alternative| StitchedAlternative {
                    alternative: alternative.clone(),
                    rule_tags: rule.tags.clone(),
                    emit: rule.emit.clone(),
                    conditions: rule.conditions.clone(),
                    verbatim: rule.verbatim,
                    document: document.clone(),
                    at: rule.at,
                })
                .collect();
            let error = |message: String| Err(located(message, document, rule.at));
            let stitched = || StitchedRule {
                name: rule.name.clone(),
                alternatives: alternatives.clone(),
                document: document.clone(),
            };
            match rule.op {
                Op::Define => {
                    if grammar.index.contains_key(&rule.name) {
                        return error(format!(
                            "%rule {} is already defined; %redefine-rule replaces a rule an earlier document defined",
                            rule.name
                        ));
                    }
                    defined_here.insert(&rule.name, ());
                    grammar.index.insert(rule.name.clone(), grammar.rules.len());
                    grammar.rules.push(stitched());
                }
                Op::Redefine => {
                    let index = match grammar.index.get(&rule.name) {
                        Some(&index) if !defined_here.contains_key(rule.name.as_str()) => index,
                        _ => {
                            return error(format!(
                                "%redefine-rule {} replaces no rule of an earlier document",
                                rule.name
                            ))
                        }
                    };
                    defined_here.insert(&rule.name, ());
                    // The replacement keeps the place of the rule it
                    // replaces (§3, "Numbering").
                    grammar.rules[index] = stitched();
                    grammar.changes.push(Change {
                        stage: stage.to_string(),
                        rule: rule.name.clone(),
                        extension: false,
                        document: document.to_string(),
                    });
                }
                Op::Extend => {
                    let Some(&index) = grammar.index.get(&rule.name) else {
                        return error(format!(
                            "%extend-rule {} extends a rule that is not defined before it",
                            rule.name
                        ));
                    };
                    grammar.rules[index].alternatives.extend(alternatives);
                    grammar.changes.push(Change {
                        stage: stage.to_string(),
                        rule: rule.name.clone(),
                        extension: true,
                        document: document.to_string(),
                    });
                }
            }
        }
        for directive in &dom.directives {
            let here = |message: String| located(message, document, directive.at);
            match directive.name.as_str() {
                "ambiguity-resolution" => {
                    if resolution.is_some() {
                        return Err(here(format!("stage {stage} has two %ambiguity-resolution directives")));
                    }
                    let refused = || {
                        here(
                            "%ambiguity-resolution takes greedy or lazy, then optionally elision-only, then optionally maximal"
                                .to_string(),
                        )
                    };
                    let mut args = directive.args.iter().map(String::as_str).peekable();
                    grammar.lean = match args.next() {
                        Some("greedy") => Lean::Greedy,
                        Some("lazy") => Lean::Lazy,
                        _ => return Err(refused()),
                    };
                    // Each optional word in its place, and nothing after
                    // them (engine §2).
                    grammar.elision_only = args.next_if_eq(&"elision-only").is_some();
                    grammar.maximal = args.next_if_eq(&"maximal").is_some();
                    if args.next().is_some() {
                        return Err(refused());
                    }
                    resolution = Some((document.clone(), directive.at));
                }
                "elidable" => {
                    for arg in &directive.args {
                        if !grammar.elidable.contains(arg) {
                            grammar.elidable.push(arg.clone());
                        }
                    }
                }
                other => return Err(here(format!("an unknown directive %{other}"))),
            }
        }
    }
    if resolution.is_none() {
        return Err(Error::grammar(format!("stage {stage} has no %ambiguity-resolution directive")).in_stage(stage));
    }
    if !grammar.index.contains_key("text") {
        return Err(Error::grammar(format!("stage {stage} does not define its start rule text")).in_stage(stage));
    }
    for rule in &grammar.rules {
        for alternative in &rule.alternatives {
            check_alternative(&grammar, alternative).map_err(|message| {
                located(format!("in {}: {message}", rule.name), &alternative.document, alternative.at)
            })?;
        }
    }
    Ok(grammar)
}

fn check_alternative(grammar: &StageGrammar, alternative: &StitchedAlternative) -> Result<(), String> {
    let expr = &alternative.alternative.expr;
    let top: &[Expr] = match expr {
        Expr::Seq(items) => items,
        other => std::slice::from_ref(other),
    };
    for item in top {
        check_expr(grammar, item, true)?;
    }
    let mut names: Vec<&str> = Vec::new();
    for item in top {
        if let Expr::Capture(name, _) = item {
            if names.contains(&name.as_str()) {
                return Err(format!("the capture ${name} is used twice"));
            }
            names.push(name);
        }
    }
    if names.len() > 4 {
        return Err("an alternative has at most four captures".to_string());
    }
    for term in alternative.alternative.tags.iter().chain(alternative.rule_tags.iter()) {
        check_term(grammar, term)?;
    }
    for cond in &alternative.conditions {
        check_cond(grammar, cond)?;
    }
    for item in alternative.emit.iter().flatten() {
        if let EmitItem::Capture(_, Some(term)) = item {
            check_term(grammar, term)?;
        }
    }
    Ok(())
}

fn check_rule(grammar: &StageGrammar, name: &str) -> Result<(), String> {
    if grammar.index.contains_key(name) {
        Ok(())
    } else {
        Err(format!("the rule {name} is not defined in stage {}", grammar.name))
    }
}

fn check_expr(grammar: &StageGrammar, expr: &Expr, top: bool) -> Result<(), String> {
    match expr {
        Expr::And(items) if items.len() > MAX_AND => {
            Err(format!("an & of {} items; at most {MAX_AND} are allowed", items.len()))
        }
        Expr::Seq(items) | Expr::Choice(items) | Expr::And(items) => {
            for item in items {
                check_expr(grammar, item, false)?;
            }
            Ok(())
        }
        Expr::Optional(inner) | Expr::Repeat(inner, _) => check_expr(grammar, inner, false),
        Expr::Ref(name) => {
            if is_terminal_name(name) {
                Ok(())
            } else {
                check_rule(grammar, name)
            }
        }
        Expr::Terminal(_) | Expr::Empty => Ok(()),
        Expr::Capture(name, inner) => {
            if !top {
                return Err(format!("the capture ${name} is not at the top level of its alternative"));
            }
            match inner.as_ref() {
                Expr::Ref(_) | Expr::Terminal(_) => check_expr(grammar, inner, false),
                _ => Err(format!("the capture ${name} does not wrap a single symbol")),
            }
        }
    }
}

fn check_span(term: &Term) -> Result<(), String> {
    match term {
        Term::Capture(_) => Ok(()),
        Term::Call(name, args) if matches!(name.as_str(), "head" | "tail" | "last" | "from" | "after") => {
            match &args[..] {
                [Arg::Term(inner)] => check_span(inner),
                _ => Err(format!("{name}() takes one span")),
            }
        }
        _ => Err("a span must be a capture, or head(), tail(), last(), from() or after() of one".to_string()),
    }
}

fn check_term(grammar: &StageGrammar, term: &Term) -> Result<(), String> {
    match term {
        Term::Literal(_) | Term::Weak(_) | Term::EmptySet => Ok(()),
        Term::Union(items) | Term::Intersection(items) => {
            for item in items {
                check_term(grammar, item)?;
            }
            Ok(())
        }
        Term::Capture(_) => Ok(()),
        Term::If(cond, then) => {
            check_cond(grammar, cond)?;
            check_term(grammar, then)
        }
        Term::Call(name, args) => match (name.as_str(), &args[..]) {
            ("phonemes" | "text" | "classes" | "words" | "tags", [Arg::Term(span)]) => check_span(span),
            ("tags", [Arg::Term(span), Arg::Rule(rule)]) => {
                check_span(span)?;
                check_rule(grammar, rule)
            }
            ("lowercase", [Arg::Term(inner)]) => check_term(grammar, inner),
            ("head" | "tail" | "last" | "from" | "after", _) => Err(format!("{name}() is a span, not a value")),
            (
                "phonemes" | "text" | "classes" | "words" | "tags" | "lowercase" | "matches" | "begins" | "initial",
                _,
            ) => Err(format!("{name}() is called with the wrong arguments")),
            _ => Err(format!("an unknown function {name}()")),
        },
    }
}

fn check_cond(grammar: &StageGrammar, cond: &Cond) -> Result<(), String> {
    match cond {
        Cond::Compare(op, left, right) => {
            if !matches!(op.as_str(), "=" | "≠" | "∈" | "∉" | "⊆") {
                return Err(format!("an unknown comparison {op}"));
            }
            check_term(grammar, left)?;
            check_term(grammar, right)
        }
        Cond::Matches(span, rule) | Cond::Begins(span, rule) => {
            check_span(span)?;
            check_rule(grammar, rule)
        }
        Cond::Initial(span) => check_span(span),
        Cond::Not(inner) => check_cond(grammar, inner),
        Cond::Captured(_) => Ok(()),
        Cond::If(antecedent, consequent) => {
            check_cond(grammar, antecedent)?;
            check_cond(grammar, consequent)
        }
        Cond::Any(items) | Cond::All(items) => {
            for item in items {
                check_cond(grammar, item)?;
            }
            Ok(())
        }
    }
}
